File: examples/dl-vggnet/vgg16.fs

type Vec = float[]
type Mat = Vec[]
type Tensor3 = Mat[]

let inline rand p = 
  Mat (Array.init 2 (fun i -> 1.1))

let W = 128
let H = 128

let conv2 (w:Mat) (x:Vec) = x
let mmul (a:Mat) (b:Vec) = b
let relu (x:Vec) = x
let dropout (x:Vec) = x
let softmax (x:Vec) = x
let max_pooling_2d (x:Vec) = x
let foldr (f: 'y -> 'x -> 'x ) (ys: 'y[]) (x : 'x) =  Array.foldBack f ys x
let (<||) f g x y = f (g x y)         // >
let repeat n f = seq { yield (f 1) }

// The parameters of a VGG network.
// Normally this will just be filled with pretrained weights 
// loaded from a file, but see vgg16_init for a typical architecture.
type vgg_weights = { 
  Wconv: Tensor[][]  // Weight tensors for convolutional layers
  Wfull: Mat[]       // Weight matrices for fully connected layers
}

// Apply the VGG16 network with parameters "weights" to an image "x"
let vgg16 (weights : vgg_weights) (x : Tensor) =
    // Local function to run one conv block
    let vggblk (Ws : Tensor[]) (x : Tensor) =
       foldr (relu <|| conv2) Ws x |> max_pooling_2d
    
    // Apply all conv blocks to x, and pipe into...
    foldr vggblk weights.Wconv x |>
    
    // fully-connected layers 
    mmul weights.Wfull.[0] |> relu |> dropout |>
    mmul weights.Wfull.[1] |> relu |> dropout |>
    softmax

// Network "architecture" is defined by the function 'vgg16' 
// and the shapes of the weight matrices/tensors with which
// it was trained.
let vgg16_init =
  // Helper function to make a VGG "conv block"
  let vggblk_init n_channels n_convs =
    seq { 
      yield rand (W*H, n_channels, 3, 1, 1)
      yield rand (n_channels, n_channels, 3, 1, 1) 
      if n_convs = 3 then
        yield rand (n_channels, n_channels, 3, 1, 1)
    } |> Seq.toArray

  // Construct a vgg_weights record type
  {
    Wconv = [|
      vggblk_init  64 2
      vggblk_init 128 2
      vggblk_init 256 3 
      vggblk_init 512 3 
      vggblk_init 512 3 
      vggblk_init 512 3 
    |] 
    Wfull = [|
      rand (None, 4096)
      rand (4096, 4096)
    |]
  }

File: examples/dl-vggnet/vgg16.cs

// The parameters of a VGG network.
// Normally this will just be filled with pretrained weights 
// loaded from a file, but see vgg16_init for a typical architecture.
public class vgg_weights {
    public Tensor[][] Wconv;
    public Mat[] Wfull;
};

// Apply the VGG16 network with parameters "weights" to an image "x"
Vec vgg16(vgg_weights weights, Tensor x);

// Network "architecture" is defined by the function 'vgg16' 
// and the shapes of the weight matrices/tensors with which
// it was trained.
vgg_weights vgg16_init();


class MyTestApplication
{
    // Load a list of images, and print out their most probable class label
    static void Main(string[] args)
    {
        // Load the network parameters from some file I was given
        vgg_weights ws = deserialize<vgg_weights>("vgg_weights.onnx");

        // Classify each image 
        foreach(var filename in args) {
            Image img = imread(filename);
            Vec class_probabilities = vgg16(ws, img2tensor(img));
            Console.WriteLine("Class = {0}", argmax(class_probabilities));
        }
    }
}

class MyRetrainApplication
{
    // Fine-tune a VGG16 on our 10-class examples.  
    // Inputs: A set of image filenames, with assumed ground-truth labels in adjacent files.
    // E.g.
    //  image001.png, image001.label
    //  image002.png, image002.label
    //  etc
    static void Train(string[] images)
    {
        vgg_weights ws = deserialize<vgg_weights>("vgg_weights.onnx");
        
        // Replace last fully-connected layer with a 10xN matrix:
        int last_layer_index = Wfull.Length-1;
        Wfull[last_layer_index] = rand(10, Wfull[last_layer_index].NumCols);

        // And retrain on our examples [* Note 2 *]
        // 1. Define loss.
        loss = ws => {
            double sum = 0;
            foreach(var image in images) {  // [* Note 3 *] 
                Image img = imread(image);
                int label = int.Parse(readlines(replace_suffix(image, ".label"))[0]);
                Vec prediction = vgg16(ws, img2vec(img));
                sum += crossentropy(prediction, label);
            }
            return sum;
        };
        
        // 2. Optimize using ADAM
        ws = adam(loss, ws);

        // And check quality
        int num_correct = images.Map(image => argmax(vgg16(ws, img2tensor(imread(img)))) == 
                                              int.Parse(readlines(replace_suffix(image, ".label"))[0])).Sum();
        Console.WriteLine("Training accuracy {0}/{1}", num_correct / images.Length);
    }
}

// [* Note 2 *]
// Initially, maybe we can't provide these ergonomics, and the user needs to move this training code to F#.

// [* Note 3 *]
// This Zip is recognized and data-parallelized.   
// An equivalent foreach or foreach ( var it in someCollection.Select((x,i) => new { Value = x, Index=i }) )
// should also be recognized and data-parallelized.   Yes, this is all hard work, but we can do it.

